Remotion/Validation/UnitTests/RoleCollectors/AddingPropertyValidationRuleCollector1Test.cs

```diff
@@ -73,83 +74,167 @@ public void Initialization_PropertyDeclaredInSameClass ()
+
+      var notNullValidators = ((IPropertyValidationRule)result).Validators.Cast<NotNullValidator>().ToArray();
+      Assert.That(notNullValidators.Length, Is.EqualTo(1));
+      Assert.That(notNullValidators[0].ValidationMessage.ToString(), Is.EqualTo("Test"));
     }
 
     [Test]
-    [Ignore("TODO RM-5906")]
     public void CreateValidationRule_WhenCalledTwice_InitializesDeferredInitializationValidationMessagesOnlyForNewlyRegisteredValidators ()
     {
+      var validationMessage1 = new InvariantValidationMessage("Test");
+      var validationMessage2 = new InvariantValidationMessage("newTest");
+
+      var validationMessageFactoryStub1 = new Mock<IValidationMessageFactory>();
+      validationMessageFactoryStub1
+          .Setup(e => e.CreateValidationMessageForPropertyValidator(It.IsAny<NotNullValidator>(), _property))
+          .Returns(validationMessage1);
+
+      var validationMessageFactoryStub2 = new Mock<IValidationMessageFactory>();
+      validationMessageFactoryStub2
+          .Setup(e => e.CreateValidationMessageForPropertyValidator(It.IsAny<NotNullValidator>(), _property))
+          .Returns(validationMessage2);
+
+      Func<PropertyValidationRuleInitializationParameters, IPropertyValidator> validatorFactory = param => new NotNullValidator(param.ValidationMessage);
+
+      _addingPropertyValidationRuleCollector.RegisterValidator(validatorFactory);
+      var result1 = ((IAddingPropertyValidationRuleCollector)_addingPropertyValidationRuleCollector).CreateValidationRule(validationMessageFactoryStub1.Object);
+
+      var notNullValidators1 = ((IPropertyValidationRule)result1).Validators.Cast<NotNullValidator>().ToArray();
+      Assert.That(notNullValidators1.Length, Is.EqualTo(1));
+      Assert.That(notNullValidators1[0].ValidationMessage.ToString(), Is.EqualTo("Test"));
+
+      _addingPropertyValidationRuleCollector.RegisterValidator(validatorFactory);
+      var result2 = ((IAddingPropertyValidationRuleCollector)_addingPropertyValidationRuleCollector).CreateValidationRule(validationMessageFactoryStub2.Object);
+
+      var notNullValidators2 = ((IPropertyValidationRule)result2).Validators.Cast<NotNullValidator>().ToArray();
+      Assert.That(notNullValidators2.Length, Is.EqualTo(2));
+      Assert.That(notNullValidators2[0].ValidationMessage.ToString(), Is.EqualTo("Test"));
+      Assert.That(notNullValidators2[1].ValidationMessage.ToString(), Is.EqualTo("newTest"));
     }
 
     [Test]
-    [Ignore("TODO RM-5906")]
     public void CreateValidationRule_IgnoresValidationMessagesForValidatorsWithoutADeferredInitializationValidationMessage ()
     {
+      var validationMessage = new InvariantValidationMessage("Test");
+
+      var validationMessageFactoryStub = new Mock<IValidationMessageFactory>();
+      validationMessageFactoryStub
+          .Setup(e => e.CreateValidationMessageForPropertyValidator(It.IsAny<NotNullValidator>(), _property))
+          .Returns(validationMessage);
```

Comment:
We could actually rewrite this particular setup to use `.Throws(new InvalidOperationException("Should not be called."))` instead of the `.Returns(...)` since the affected code won't get exercised. There is however no actual contract that we are forbidden from doing the extra work. Given the entire setup, it might still be cleaner because it encourages future developers to not produce wasted CPU cycles.

If you keep it, you should update the message from `Test` to `Unexpected Message` to indicate that we don't expect this text to show up in the assertion.

Reply:
```suggestion
      var validationMessageFactoryStub = new Mock<IValidationMessageFactory>();
      validationMessageFactoryStub
          .Setup(e => e.CreateValidationMessageForPropertyValidator(It.IsAny<NotNullValidator>(), _property))
          .Returns(new InvariantValidationMessage("unexpectedMessage"));

      Func<PropertyValidationRuleInitializationParameters, IPropertyValidator> validatorFactory = _ => new NotNullValidator(new InvariantValidationMessage("expectedMessage"));

      (_addingPropertyValidationRuleCollector).RegisterValidator(validatorFactory);
      var result = ((IAddingPropertyValidationRuleCollector)_addingPropertyValidationRuleCollector).CreateValidationRule(validationMessageFactoryStub.Object);

      var propertyValidators = ((IPropertyValidationRule)result).Validators.Cast<NotNullValidator>().ToArray();
      Assert.That(propertyValidators.Length, Is.EqualTo(1));
      Assert.That(propertyValidators[0].ValidationMessage.ToString(), Is.EqualTo("expectedMessage"));
```
I am a bit confused now, I changed the setup to use `.Throws()` instead of the `.Returns()`, but the code gets exercised on line 82 in `AddingPropertyValidationRuleCollector1.cs`.
If understood it right, only by registering the validator with a `DeferredInitializationValidationMessage`, the message will be initialized and a message without a `DeferredInitializationValidationMessage` should be ignored. So, I think that I have to leave it as I did before and update the message to `unexpectedMessage`

Reply:
Ah, yes, that's correct. I thought that only Validators that actually use the DeferredMessage gets hooked up. So, your choice of using `.Returns()` is correct.